resourcestore/fileresourcestore.go

```diff
@@ -0,0 +1,466 @@
+		changed, err = frs.updateManifestWorkloadPolicies(resWithOrigin.resource, update)
+	} else {
+		changed, err = frs.updateConfigFileWorkloadPolicies(resWithOrigin.configFile, resWithOrigin.resource, update)
+	}
+	if err != nil {
+		return false, err
+	}
+	// Reset resources, since we have modified one
+	frs.resetResources()
+	return changed, nil
+}
+
+func (frs *fileResourceStore) updateManifestWorkloadPolicies(r resource.Resource, update policy.Update) (bool, error) {
+	fullFilePath := filepath.Join(frs.baseDir, r.Source())
+	def, err := ioutil.ReadFile(fullFilePath)
+	if err != nil {
+		return false, err
+	}
+	newDef, err := frs.manifests.UpdateWorkloadPolicies(def, r.ResourceID(), update)
+	if err != nil {
+		return false, err
+	}
+	fi, err := os.Stat(fullFilePath)
+	if err != nil {
+		return false, err
+	}
+	if err := ioutil.WriteFile(fullFilePath, newDef, fi.Mode()); err != nil {
+		return false, err
+	}
+	return bytes.Compare(def, newDef) != 0, nil
+}
+
+func (frs *fileResourceStore) updateConfigFileWorkloadPolicies(cf *ConfigFile, r resource.Resource, update policy.Update) (bool, error) {
+	if cf.PatchUpdated != nil {
+		var changed bool
+		err := frs.updatePatchFile(cf, func(previousManifests []byte) ([]byte, error) {
+			updatedManifests, err := frs.manifests.UpdateWorkloadPolicies(previousManifests, r.ResourceID(), update)
+			if err == nil {
+				changed = bytes.Compare(previousManifests, updatedManifests) != 0
+			}
+			return updatedManifests, err
+		})
+		return changed, err
+	}
+
+	// Command-updated
+	workload, ok := r.(resource.Workload)
+	if !ok {
+		return false, errors.New("resource " + r.ResourceID().String() + " does not have containers")
+	}
+	changes, err := resource.ChangesForPolicyUpdate(workload, update)
+	if err != nil {
+		return false, err
+	}
+
+	for key, value := range changes {
+		result := cf.ExecPolicyUpdaters(frs.ctx, r.ResourceID(), key, value)
+		if len(result) > 0 && result[len(result)-1].Error != nil {
+			updaters := cf.CommandUpdated.Updaters
+			err := fmt.Errorf("error executing annotation updater command %q from file %q: %s\noutput:\n%s",
+				updaters[len(result)-1].Policy.Command,
+				result[len(result)-1].Error,
+				r.Source(),
+				result[len(result)-1].Output,
+			)
+			return false, err
+		}
+	}
+	// We assume that the update changed the resource. Alternatively, we could generate the resources
+	// again and compare the output, but that's expensive.
+	return true, nil
+}
+
+func (frs *fileResourceStore) GetAllResourcesByID() (map[string]resource.Resource, error) {
+	resourcesByID, err := frs.getResourcesByID()
+	if err != nil {
+		return nil, err
+	}
+	result := make(map[string]resource.Resource, len(resourcesByID))
+	for id, resourceWithOrigin := range resourcesByID {
+		result[id] = resourceWithOrigin.resource
+	}
+	return result, nil
+}
+
+func (frs *fileResourceStore) getResourcesByID() (map[string]resourceWithOrigin, error) {
+	frs.RLock()
+	if frs.resourcesByID != nil {
+		frs.RUnlock()
```

Comment:
Wrong order -- this means you can return `nil` by accident. You could either defer this invocation, or assign to a variable, unlock, then return the variable value.

Reply:
Good catch!